### ISSUE TITLE
CloneMethod.freeze

### ISSUE THREAD

**Issue by rems75** [Microsoft/CNTK#2196]
Hello,
I'm using the code from CNTK/Examples/ReinforcementLearning/DeepQNeuralNetwork.py to do a few experiments.
And surprisingly the CloneMethod.freeze does not seem to work. Evaluating the target network self._target_net and the "real" one self._action_value_net on the states during training gives the exact same q values at each time step.

Second remark, adding unit_gain=False to the adam learner makes everything diverge.

Any help is very welcome.

Best,
Remi

**Comment by mfuntowicz** [comment]
Hi @rems75,

Regarding Adam learner, it might not handle correctly the unit_gain parameter, Nikos might be able to agree or not on this point. When using Adam, I would argue to stay on the unit_gain=True parameter for now.

Regarding the Clone method how did you evaluate the output of both network ? 

The target network is updated from the action_value one every (_by default_) 40k steps. If you look at the values of both network directly after an update, values might be very very similar, where looking at value when the target network is about to be updated might give you different results (after 39k steps for example).

Morgan

**Comment by rems75** [comment]
Hi @mfuntowicz 

I'm in touch with Nikos about the unit_gain question. The use_mean_gradient parameter is also a big difference between cntk and other frameworks.

Upgrading to CNTK 2.1 solved the issue with the target network. CloneMethod.freeze does not work on CNTK 2.0. 

Cheers,
Remi

**Comment by ebarsoum** [comment]
If you set unit_gain=False, you will need to reduce your learning rate. use_mean_gradient if it is set to True, it will match the behavior of other framework. CNTK by default sum the gradient in the MB, by setting 'use_mean_gradient' we will average them. To match other framework behavior you need to set both flags.

**Comment by rems75** [comment]
Hi @ebarsoum ,

Thanks for the clarification. 
I am actually quite interested in understanding a surprising behavior of the original algorithm (with all the flags set by default). The target network was disabled, but the agent still managed to learn good q-values and an optimal policy. 
What exactly does unit_gain do in the case of the Adam optimizer? There is already a normalization by (1-momentum) in the original algorithm.

**Comment by n17s** [comment]
`unit_gain=True` does exactly what is done in the Adam paper. For uniformity with the other learners we also added a `unit_gain=False` that replaces 1-beta_1 in the algorithm with 1.

Closing this as the original questions has been addressed.  Feel free to open new issues about other topics.
